### PR TITLE
Gradle 8.4

### LINKED ISSUE #12655: Upgrade to Gradle 8.4
### Description

Upgrade Gradle from 7.6 to 8.4 - supports building directly with JDK 21 LTS.

https://docs.gradle.org/8.4/release-notes.html

### PR DESCRIPTION
**Pull request by risdenk**
Fixes https://github.com/apache/lucene/issues/12655

### Description

Upgrades Gradle from 7.6 to 8.4 - supports building directly with JDK 21 LTS.
* https://docs.gradle.org/8.4/release-notes.html

Upgrades a few build plugins to support JDK 21 as well
* `net.ltgt.errorprone` - 3.0.1 -> 3.1.0
* `googleJavaFormat` - 1.15.0 -> 1.18.1 - this caused tidy to be run and regen for quite a few things but its mostly just whitespace changes

Also fixed a python 3.9 or later issue with gen_ForUtil.py - https://bugs.python.org/issue39350 - gcd was moved from `fractions` to `math` in Python 3.5 and removed from `fractions` in python 3.9.

I had been looking at Gradle 8.x in Solr - https://github.com/apache/solr/pull/1997

### CODE REVIEW

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene84/gen_ForUtil.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from fractions import gcd
+from math import gcd
```
Comment (risdenk) [inline lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene84/gen_ForUtil.py:18]: Python 3.5 added `gcd` to `math` and Python 3.9 removed from `fractions` - https://bugs.python.org/issue39350

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/gen_ForUtil.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from fractions import gcd
+from math import gcd
```
Comment (risdenk) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/gen_ForUtil.py:18]: Python 3.5 added `gcd` to `math` and Python 3.9 removed from `fractions` - https://bugs.python.org/issue39350

**Comment by dweiss** [comment]
Oh, impressive work. I'm surprised it's worked because I know there are incompatibilities between palantir's consistent version plugin and newer spotless releases.

Overall, it looks great. I would strongly encourage to rewrite patch history a bit though - separate all build-related changes from the google format upgrade. Then apply build changes (one patch), apply google format upgrade (second patch), commit auto-reformatted/ regenerated code (third patch), add the third commit SHA to .git-blame-ignore-revs.

**Comment by dweiss** [comment]
Just to be clear - the above is not _strictly_ necessary but I think it'd make for a much saner commit history and ability to review what you did to the build files, otherwise it gets obscured by auto-formatting diffs. The suggested changes can come as a single PR, but then should be merged using a merge commit (so that the history of individual commits is there).

File: gradlew [hunk 3]
```diff
@@ -1,7 +1,7 @@
-#!/usr/bin/env sh
```
Comment (dweiss) [inline gradlew:1]: this was intentional (/usr/bin/env).

**Comment by uschindler** [comment]
> Oh, impressive work. I'm surprised it's worked because I know there are incompatibilities between palantir's consistent version plugin and newer spotless releases.
> 
> Overall, it looks great. I would strongly encourage to rewrite patch history a bit though - separate all build-related changes from the google format upgrade. Then apply build changes (one patch), apply google format upgrade (second patch), commit auto-reformatted/ regenerated code (third patch), add the third commit SHA to .git-blame-ignore-revs.

I wanted to say the same thing. Maybe rewrite the changes on this PR as only 2 commits. And then merge the PR without squashing.

**Comment by risdenk** [comment]
I pushed changes based on @dweiss suggestion of

> separate all build-related changes from the google format upgrade. Then apply build changes (one patch), apply google format upgrade (second patch), commit auto-reformatted/ regenerated code (third patch), add the third commit SHA to .git-blame-ignore-revs.

However `add the third commit SHA to .git-blame-ignore-revs.` will probably have to wait until this is merged into main/branch_9x to make sure the right commit SHA ends up in the file.

**Review by uschindler — APPROVED** [review]
Hi,
this looks fine to me. Unfortunately, the defaults in Github disallow merge requests and enforce you to sqash. So I'd suggest to not use Github PR functionality and merge manually using GIT client and close the PR afterwards.

In additon, on 9.x you should only cherry-pick the first 2 commits and then apply code formatting separately and commit this as a separate 9.x commit and add that revision to the list of ignored revs for blame in 9x.

**Comment by risdenk** [comment]
Agreed @uschindler I created https://github.com/apache/lucene/issues/12655 so I can prefix each commit w/ the issue identifier.

**Comment by risdenk** [comment]
Merged to main as:
* 30d3eba93314fbaa014e885a21d2aa1588433df5
* 2c42b8941aa3247b8e24451af1553993f6a95079
* de3b294be44c2cea51a5e909db05043aa8d15150
* a8fba38f1691199e0813d71783ecda444c1d8ca3

**Comment by uschindler** [comment]
Thanks! 👍

**Pull request closed without merging**